aten/src/THC/THCOffsetInfo.cuh

```diff
@@ -0,0 +1,274 @@
+// "carries over" to #k-1.  In general, carryDelta can be "negative" even though
+// IndexType is unsigned: it is always added to the offset, so the result will
+// be correct.)
+//
+// So, the example shown above will be stored as:
+//
+//      increments  = (1, 32, 4)
+//      scaledIncrs = (2000, 640, 4)
+//      carryDelta  = (_, 600, 10)
+//
+// Finally, we can also optimize the initial coordinate computation by
+// pre-computing "magic dividers", thus replacing the initial div/mod operations
+// by multiplication.  (See IntDivider for details.)
+
+// Helper function that increments 'indices' and returns the updated offset.
+template <typename IndexType>
+__host__ __device__ __forceinline__
+IndexType incrementIdx(int dims, IndexType offset,
+                       const IntDivider<IndexType> sizes[],
+                       const IndexType increments[],
+                       const IndexType scaledIncrs[],
+                       const IndexType carryDelta[],
+                       IndexType indices[])
+{
+  bool carry = false;
+
+  for (int i = dims - 1; i > 0; --i) {
+    IndexType index = indices[i] + increments[i] + (IndexType) carry;
+    offset += scaledIncrs[i];
+
+    // Note that 'index' here may reach (2 * sizes[i] - 1), because indices[i]
+    // and increments[i] can each be up to (sizes[i] - 1), and there can be a
+    // carry.  Assuming 32-bit indices, sizes[i] can be at most UINT32_MAX =
+    // 0xffffffff (otherwise we will be using 64-bit indices), so the "true
+    // value" of 'index' can be at most 0x1fffffffd, larger than UINT32_MAX.
+    //
+    // The second condition below checks if 'index' had an overflow.
+    carry = (index >= sizes[i].divisor) || (index < indices[i]);
+    if (carry) {
+      index -= sizes[i].divisor;
+      offset += carryDelta[i];
+    }
+
+    indices[i] = index;
+  }
+
+  // Dimension 0 is special because we are guaranteed to stay within bound.
+  offset += scaledIncrs[0];
+
+  return offset;
+}
+
+// Helper class that keeps track of the linear index.
+template <typename IndexType> struct LinearIdIterator {
+  // NOTE: The typecasting in 'index' and 'step' is essential when IndexType is
+  //       64-bit; without casting, results are silently truncated to 32 bits!
+  __device__ explicit LinearIdIterator(IndexType limit)
+    : index((IndexType) blockIdx.x * blockDim.x + threadIdx.x),
+      step((IndexType) gridDim.x * blockDim.x),
+      limit(limit), hasNext(index < limit) { }
+
+  // Explicit version for testing.
+  __host__ __device__
+  LinearIdIterator(IndexType start, IndexType step, IndexType limit)
+    : index(start), step(step), limit(limit), hasNext(start < limit) { }
+
+  // Assumes 'hasNext' is true before called.
+  __host__ __device__ void increment() {
+    IndexType next = index + step;
+
+    // The second condition is necessary to handle overflow (e.g., when step is
+    // 2GB and limit is 3GB, assuming 32-bit index).
```

Comment:
In the case where the step is 2 GB and the limit is 3 GB, why is the index 32 bits?

If this is the case, this feels like a bug in the code used to determine whether or not 64 bit indexing math is needed. This particular check shouldn't be needed, as it should be guarded against before we even get here.

Reply:
You can see the function here:

https://github.com/pytorch/pytorch/blob/78649419c42c157db1c624df5e5020a78b007dd6/aten/src/THC/THCTensorTypeUtils.cuh#L75

https://github.com/pytorch/pytorch/blob/78649419c42c157db1c624df5e5020a78b007dd6/aten/src/THC/THCTensorTypeUtils.cu#L247

It basically admits any tensor with size up to `UINT32_MAX - 1`.  Shall I change it to instead accept up to `INT32MAX - 1` ?  (On one hand, tensors with 2-4G elements will now have to use 64-bit math.  On the other hand, given that the code basically didn't work before, I think nobody will be affected by the slowdown.)

Reply:
Changing it to INT32MAX - 1 is fine, people would hardly use such large sizes and this is just to make sure it still executes correctly.

Reply:
Done.